### DISCUSSION TITLE
SourceGenerators as a replacement for VS extensions?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/roslyn#50255] [Q&A] [answered]
This is just question for SourceGenerators for some items that were not quite clear from the current design specs.

We currently have a VS extension that is invoked in a single CUDA (*.cu, e.g. FFT.cu) file every time the user saves this files.
Internally on saving, the extension invokes the NVCC compiler, deliver by nvidia and creates a so called PTX file (here: FFT.ptx) which is added to the project if not already available (if already available it will be overriden). It is also marked as embedded resource (related to #49923).

Additionally, we run clang to extract the syntax tree from the CUDA code (from FFT.cu) and created a wrapper for the CUDA code so that it can be invoked from C#. This wrapper also loads the embedded resources (here: FFT.ptx) and does the necessary CUDA things. This wrapper is created or added to the project (here: FFT.designer.cs), marked with "compile". As above if it already there, it will be overriden.

Addtionally, another file for user specific code is created (here: FFT.cs). But only once, it is not overriden when the users saves the CUDA code (FFT.cu).

There reason why this cannot be done during typing is that the process takes a while (some 5-10 seconds) and therefore we have decided only to do it when saving the code.

One problem is that extensions cannot be versioned and if I want to have another v ersion of the extension and need to uninstall it and install another version. So if I want to go back in source code history and in the meantime the CUDA extension has changed and need to know about this, otherwise I would create wrong ptx code. Also debugging extensions is a nightmare.

So in my opinion source generator would be the optimal tool. I can add it via nuget, which versions itself automatically and I can add this solution wide via Directory.Build.props.

My question is can source generator currently do this, and if not will they be able to do it in future versions?

**Comment by jasonmalinowski** [comment, marked as answer]
_Today_ you'll run into a few problems. A source generator is capable of producing .cs files, so your wrapper could be easily produced. As you observed though we don't support generating embedded resources nor non-.cs files, so for now you wouldn't be able to do that. We're currently working on designs though for that so we absolutely expect those should be doable in future versions.

> There reason why this cannot be done during typing is that the process takes a while (some 5-10 seconds)

What parts of the process you describe are the slow ones? At this point we don't have a great plan for generators that take a _really_ long time like this, but that's something further we'd have to work on. Better understanding your use case here would help us accommodate that scenario better.

**Reply by msedi** [reply]
@jasonmalinowski. What be awesome the things described above would work with source generators, although I understand that the it's maybe not the original intended use of source generators.

> What parts of the process you describe are the slow ones

In the end I have to call an external process (nvcc.exe) that generates the ptx code (which can be slow). Additionally I have to call clang.exe (or at least the API) to generate the syntax tree to creates the necessary C# wrappers. This process takes some time and does maybe also not fit into the intended use of source generators.

The question only came up since debugging and creating VS extensions is a bit of a pain, where source generators seem to fit more our needs (especially when it comes to the question of having a versioning).

**Reply by jasonmalinowski** [reply]
So we absolutely imagine these sorts of cases being done by generators, so you're not off there. Is the clang parsing to generate the wrapper at least fast? Because it might be useful here to have a better split so the generator can quickly rewrite the C# wrapper (so you have live intellisense in the IDE) but the slower nvcc.exe invocation only happens on actual build.

Now that's one thing to think about too: if you're using source generators it then means you're going to be running nvcc.exe on every build. On one hand that might be what you want, on the other hand it might slow your build down where at least today you have the _slight_ advantage that it's being done up front by the developer, right? The ptx files get checked in to source control?

**Reply by msedi** [reply]
That's nice to hear ;-) I'm not quite sure right know, but the clang part is not such a big problem. But maybe not so fast that it fits in into the regular roslyn chain. 

I read something in your specs that there will be a possibility to register for IDE events. But the more I think about it I'm not quite sure if the source generator should be "misused" for this, or if there should be something different. As you said it will be created on every build where the source generator could kill the nvcc process and restart it when typing but I assume that this is really out of scope for source generator and maybe should be.

And yes, the ptx code and all the other parts are added to the source control.

While our discussion I think I need more something like an extension that can be added via nuget to the project that extends certain mechanisms of VS or something completely different. I'm not sure if the project is the right place and if you in general require a project to be independent of VS?

**Reply by msedi** [reply]
@jasonmalinowski: I checked again with visual studio. Currently we are using the custom tool and are creating our own IVsSingleFileGenerator. I think if this concept can be extended to also allow generators added to the project as nuget packages this would help. But I think this is nothing for this discussion, right?

**Reply by jasonmalinowski** [reply]
So I don't imagine you're necessarily registering for "IDE events" as much as having more granular ways to break apart the different parts of things that run. The IDE needs the C# source files for your inner editing loop, but we don't need the embedded resources until you're doing a build, or potentially edit-and-continue.

**Reply by jasonmalinowski** [reply]
> I think if this concept can be extended to also allow generators added to the project as nuget packages this would help. But I think this is nothing for this discussion, right?

I guess we imagine source generators as our replacement, and we'll continue to beef up the design as needed.

**Reply by msedi** [reply]
@jasonmalinowski. Thanks a lot for the discussion. I will wait until you have more details/improvements on your side ;-) If you need some details of my current situation -> Just ask.

**Reply by jasonmalinowski** [reply]
@msedi I think the one specific detail would be for your clang processing, how expensive is it? Is the process on the order of a few milliseconds, tens, hundreds, or a second or more? Also, your extension today that reprocesses a CUDA file, when a single file is modified you only have to recompute _that_ file, right? You don't have to reanalyze all files at once?
